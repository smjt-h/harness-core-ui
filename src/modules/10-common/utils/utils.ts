/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}

export const isCommunityPlan = (): boolean => false

export const isOnPrem = (): boolean => window.deploymentType === 'ON_PREM'

export const focusOnNode = (node: HTMLElement): void => {
  const oldTabIndex = node.tabIndex
  node.tabIndex = -1
  node.focus()
  node.tabIndex = oldTabIndex
}

const HOTJAR_SUPPRESSION_ATTR = 'data-hj-suppress'

// Utility to add `data-hj-suppress` into a collection of elements to
// suppress data from HotJar recording
// @see https://bit.ly/3rCgpOY
export const suppressHotJarRecording = (elements: Element[] | null | undefined): void => {
  if (window.hj) {
    elements?.forEach?.((e: Element) => {
      if (!e.hasAttribute(HOTJAR_SUPPRESSION_ATTR)) {
        e.setAttribute(HOTJAR_SUPPRESSION_ATTR, 'true')
      }
    })
  }
}

// Utility to generate { 'data-hj-suppress': true } attribute if HotJar is available
export const addHotJarSuppressionAttribute = (): { [HOTJAR_SUPPRESSION_ATTR]: boolean } | undefined =>
  window.hj ? { [HOTJAR_SUPPRESSION_ATTR]: true } : undefined
