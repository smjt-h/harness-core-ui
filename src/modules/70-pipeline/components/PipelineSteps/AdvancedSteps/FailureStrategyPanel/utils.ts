/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { SetStateAction } from 'react'
import type { FormikErrors, FormikProps } from 'formik'
import { isEmpty } from 'lodash-es'
import { Intent } from '@blueprintjs/core'

import type {
  RetryFailureActionConfig,
  IgnoreFailureActionConfig,
  ManualInterventionFailureActionConfig,
  AbortFailureActionConfig,
  StepGroupFailureActionConfig,
  MarkAsSuccessFailureActionConfig,
  StageRollbackFailureActionConfig,
  FailureStrategyConfig,
  OnFailureConfig
} from 'services/cd-ng'

export type AllActions =
  | RetryFailureActionConfig
  | IgnoreFailureActionConfig
  | ManualInterventionFailureActionConfig
  | AbortFailureActionConfig
  | StepGroupFailureActionConfig
  | MarkAsSuccessFailureActionConfig
  | StageRollbackFailureActionConfig

export interface AllFailureStrategyConfig extends FailureStrategyConfig {
  onFailure: OnFailureConfig & { action: AllActions }
}

export function hasItems<T>(data?: T[]): boolean {
  return Array.isArray(data) && data.length > 0
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function findTabWithErrors(errors?: FormikErrors<any>): number {
  const fsErrors = errors?.failureStrategies

  return Array.isArray(fsErrors)
    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
      fsErrors.findIndex((err: any) => !isEmpty(err))
    : -1
}

export interface FormState {
  failureStrategies?: AllFailureStrategyConfig[]
}

export interface HandleChangeInStrategiesProps {
  formValues: FormState
  selectedStrategyNum: number
  setSelectedStrategyNum: (value: SetStateAction<number>) => void
  setFormikState: FormikProps<FormState>['setFormikState']
}

export function handleChangeInStrategies({
  formValues,
  selectedStrategyNum,
  setSelectedStrategyNum,
  setFormikState
}: HandleChangeInStrategiesProps): void {
  const fs = formValues.failureStrategies
  /* istanbul ignore else */
  if (Array.isArray(fs)) {
    /* istanbul ignore else */
    if (selectedStrategyNum >= fs.length) {
      // select the new last tab, if the last tab was deleted
      setSelectedStrategyNum(Math.max(0, fs.length - 1))
    }

    /* istanbul ignore else */
    if (fs.length === 0) {
      // reset errors when all the tabs are deleted
      setFormikState({ errors: {}, submitCount: 0 })
    }
  }
}

export function getTabIntent(i: number, selectedStrategyNum: number): Intent {
  return i === selectedStrategyNum ? Intent.PRIMARY : Intent.NONE
}
