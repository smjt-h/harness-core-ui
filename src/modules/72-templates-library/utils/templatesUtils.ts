/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { isEmpty } from 'lodash-es'
import type { UseStringsReturn } from 'framework/strings'
import { Scope } from '@common/interfaces/SecretsInterface'
import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import type { TemplateSummaryResponse } from 'services/template-ng'

export enum TemplateType {
  Step = 'Step',
  Stage = 'Stage',
  Pipeline = 'Pipeline',
  Service = 'Service',
  Infrastructure = 'Infrastructure',
  StepGroup = 'StepGroup',
  Execution = 'Execution',
  MonitoredService = 'MonitoredService'
}

export const AllTemplatesTypes = 'All'

export interface TemplateTypeOption {
  label: string
  value: string
  disabled?: boolean
}

export const getScopeBasedQueryParams = (
  { accountId, projectIdentifier, orgIdentifier }: ProjectPathProps,
  scope: Scope
) => {
  return {
    accountIdentifier: accountId,
    projectIdentifier: scope === Scope.PROJECT ? projectIdentifier : undefined,
    orgIdentifier: scope === Scope.PROJECT || scope === Scope.ORG ? orgIdentifier : undefined
  }
}

export const getAllowedTemplateTypes = (
  getString: UseStringsReturn['getString'],
  module?: string,
  isPipelineTemplateEnabled?: boolean
): TemplateTypeOption[] => {
  const AllowedTemplateTypes = [
    {
      label: getString('step'),
      value: TemplateType.Step,
      disabled: false
    },
    {
      label: getString('common.stage'),
      value: TemplateType.Stage,
      disabled: false
    },
    {
      label: getString('common.pipeline'),
      value: TemplateType.Pipeline,
      disabled: !isPipelineTemplateEnabled
    },
    {
      label: getString('service'),
      value: TemplateType.Service,
      disabled: true
    },
    {
      label: getString('infrastructureText'),
      value: TemplateType.Infrastructure,
      disabled: true
    },
    {
      label: getString('stepGroup'),
      value: TemplateType.StepGroup,
      disabled: true
    },
    {
      label: getString('executionText'),
      value: TemplateType.Execution,
      disabled: true
    }
  ]
  if (module === 'cv') {
    AllowedTemplateTypes.push({
      label: getString('connectors.cdng.monitoredService.label'),
      value: TemplateType.MonitoredService,
      disabled: false
    })
    return AllowedTemplateTypes.sort((a, b) => Number(a.disabled) - Number(b.disabled))
  }
  return AllowedTemplateTypes
}

export const getVersionLabelText = (template: TemplateSummaryResponse, getString: UseStringsReturn['getString']) => {
  return isEmpty(template.versionLabel)
    ? getString('templatesLibrary.alwaysUseStableVersion')
    : template.stableTemplate
    ? getString('templatesLibrary.stableVersion', { entity: template.versionLabel })
    : template.versionLabel
}
