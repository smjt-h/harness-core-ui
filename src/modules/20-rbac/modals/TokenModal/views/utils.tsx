/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { getReadableDateTime } from '@common/utils/dateUtils'

export const DATE_FORMAT = 'MM/DD/YYYY'

export const hasNoExpiryDate = (value?: number): boolean => {
  return getReadableDateTime(value, DATE_FORMAT) === '12/31/2099'
}

export const getSelectedExpiration = (value?: number): string => {
  if (!value) {
    return '30'
  } else if (hasNoExpiryDate(value)) {
    return '-1'
  }
  return 'custom'
}

export const getSelectedExpirationDate = (value: string): string => {
  switch (value) {
    case '30':
    case '90':
    case '180':
      return getReadableDateTime(new Date().setDate(new Date().getDate() + parseInt(value)), DATE_FORMAT)
    case '-1':
      return '12/31/2099'
    default:
      return ''
  }
}

export const getExpirationDate = (value?: number): string => {
  if (value) {
    return getReadableDateTime(value, DATE_FORMAT)
  } else {
    return getSelectedExpirationDate(getSelectedExpiration(value))
  }
}
