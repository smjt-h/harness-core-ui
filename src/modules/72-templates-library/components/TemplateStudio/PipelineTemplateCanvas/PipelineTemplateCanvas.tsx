/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { useParams } from 'react-router-dom'
import { omit, set } from 'lodash-es'
import { TemplateContext } from '@templates-library/components/TemplateStudio/TemplateContext/TemplateContext'
import type { GitQueryParams, ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import { useGlobalEventListener, useQueryParams } from '@common/hooks'
import { TemplatePipelineProvider } from '@pipeline/components/TemplatePipelineContext'
import { sanitize } from '@common/utils/JSONUtils'
import type { PipelineInfoConfig } from 'services/cd-ng'
import { PipelineContextType } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { DrawerTypes as PipelineDrawerTypes } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineActions'
import { DrawerTypes as TemplateDrawerTypes } from '@templates-library/components/TemplateStudio/TemplateContext/TemplateActions'
import { RightDrawer } from '@templates-library/components/TemplateStudio/RightDrawer/RightDrawer'
import StageBuilder from '@pipeline/components/PipelineStudio/StageBuilder/StageBuilder'
import { RightBar as PipelineStudioRightBar } from '@pipeline/components/PipelineStudio/RightBar/RightBar'
import { TemplateDrawer } from '@templates-library/components/TemplateDrawer/TemplateDrawer'

export const DefaultNewPipelineName = 'Pipeline Name'
export const DefaultNewPipelineId = 'pipeline_name'

const PIPELINE_TO_TEMPLATE_DRAWER_TYPE_MAP = {
  [PipelineDrawerTypes.PipelineVariables]: TemplateDrawerTypes.TemplateVariables,
  [PipelineDrawerTypes.TemplateInputs]: TemplateDrawerTypes.TemplateInputs
} as const

type EventDetailType = PipelineDrawerTypes.PipelineVariables | PipelineDrawerTypes.TemplateInputs

const PipelineTemplateCanvas = () => {
  const {
    state: {
      templateView: { isDrawerOpened }
    }
  } = React.useContext(TemplateContext)
  const {
    state: { template, isLoading, isUpdated, templateView },
    updateTemplate,
    updateTemplateView,
    isReadonly
  } = React.useContext(TemplateContext)
  const { accountId, projectIdentifier, orgIdentifier } = useParams<ProjectPathProps>()
  const { branch, repoIdentifier } = useQueryParams<GitQueryParams>()

  const createPipelineFromTemplate = () => ({
    ...template.spec,
    name: DefaultNewPipelineName,
    identifier: DefaultNewPipelineId
  })

  const [pipeline, setPipeline] = React.useState<PipelineInfoConfig>(createPipelineFromTemplate())

  React.useEffect(() => {
    if (!isLoading && !isUpdated) {
      setPipeline(createPipelineFromTemplate())
    }
  }, [isLoading, isUpdated])

  const onUpdatePipeline = async (pipelineConfig: PipelineInfoConfig) => {
    const processNode = omit(pipelineConfig, 'name', 'identifier', 'description', 'tags')
    sanitize(processNode, { removeEmptyArray: false, removeEmptyObject: false, removeEmptyString: false })
    set(template, 'spec', processNode)
    await updateTemplate(template)
  }

  useGlobalEventListener('OPEN_PIPELINE_TEMPLATE_RIGHT_DRAWER', event => {
    const adaptedDrawerType = PIPELINE_TO_TEMPLATE_DRAWER_TYPE_MAP[event.detail as EventDetailType]
    if (adaptedDrawerType) {
      updateTemplateView({
        ...templateView,
        isDrawerOpened: true,
        drawerData: { type: adaptedDrawerType }
      })
    }
  })

  if (pipeline) {
    return (
      <TemplatePipelineProvider
        queryParams={{ accountIdentifier: accountId, orgIdentifier, projectIdentifier, repoIdentifier, branch }}
        initialValue={pipeline as PipelineInfoConfig}
        onUpdatePipeline={onUpdatePipeline}
        contextType={PipelineContextType.PipelineTemplate}
        isReadOnly={isReadonly}
      >
        {!isDrawerOpened && <StageBuilder />}
        <PipelineStudioRightBar />
        <TemplateDrawer />
        <RightDrawer />
      </TemplatePipelineProvider>
    )
  } else {
    return <></>
  }
}

export const PipelineTemplateCanvasWithRef = React.forwardRef(PipelineTemplateCanvas)
