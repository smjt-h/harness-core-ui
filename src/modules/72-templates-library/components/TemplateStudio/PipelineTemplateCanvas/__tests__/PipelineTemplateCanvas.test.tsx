/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { act, fireEvent, render } from '@testing-library/react'
import produce from 'immer'
import { set } from 'lodash-es'
import { TestWrapper } from '@common/utils/testUtils'
import { TemplateContext } from '@templates-library/components/TemplateStudio/TemplateContext/TemplateContext'
import { TemplateType } from '@templates-library/utils/templatesUtils'
import { getTemplateContextMock } from '@templates-library/components/TemplateStudio/__tests__/stateMock'
import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { DrawerTypes } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineActions'
import { DrawerTypes as TemplateDrawerTypes } from '@templates-library/components/TemplateStudio/TemplateContext/TemplateActions'
import { PipelineTemplateCanvasWithRef } from '../PipelineTemplateCanvas'

jest.mock('@pipeline/components/PipelineStudio/StageBuilder/StageBuilder', () => () => {
  const {
    state: { pipeline },
    updatePipeline
  } = usePipelineContext()
  return (
    <div className="stage-builder-mock">
      <button
        onClick={() => {
          const updatedPipeline = produce(pipeline, draft => {
            set(draft, 'stages[0].stage.spec.serviceConfig.serviceRef', 'Pipeline')
          })
          updatePipeline(updatedPipeline)
        }}
      >
        Update Pipeline
      </button>
    </div>
  )
})

describe('<PipelineTemplateCanvas/> tests', () => {
  test('should call updateTemplate with updated template', async () => {
    const templateContext = getTemplateContextMock(TemplateType.Pipeline)
    const { container, getByRole } = render(
      <TestWrapper>
        <TemplateContext.Provider value={templateContext}>
          <PipelineTemplateCanvasWithRef />
        </TemplateContext.Provider>
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
    const updatePipelineButton = getByRole('button', { name: 'Update Pipeline' })
    await act(async () => {
      fireEvent.click(updatePipelineButton)
    })
    const updatedTemplate = produce(templateContext.state.template, draft => {
      set(draft, 'spec.stages[0].stage.spec.serviceConfig.serviceRef', 'Pipeline')
    })
    expect(templateContext.updateTemplate).toBeCalledWith(updatedTemplate)
  })

  test('should open templates variables drawer on a custom window event', async () => {
    const templateContext = getTemplateContextMock(TemplateType.Pipeline)
    render(
      <TestWrapper>
        <TemplateContext.Provider value={templateContext}>
          <PipelineTemplateCanvasWithRef />
        </TemplateContext.Provider>
      </TestWrapper>
    )
    window.dispatchEvent(new CustomEvent('OPEN_PIPELINE_TEMPLATE_RIGHT_DRAWER', { detail: DrawerTypes.TemplateInputs }))
    expect(templateContext.updateTemplateView).toBeCalledWith({
      drawerData: { type: TemplateDrawerTypes.TemplateInputs },
      isDrawerOpened: true,
      isYamlEditable: false
    })
  })
})
