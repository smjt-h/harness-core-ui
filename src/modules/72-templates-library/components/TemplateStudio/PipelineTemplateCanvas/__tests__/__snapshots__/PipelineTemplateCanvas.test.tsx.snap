// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PipelineTemplateCanvas/> tests should match snapshot 1`] = `
<div>
  <div
    class="stage-builder-mock"
  />
  <div
    class="rightBar"
  >
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--withLeftIcon Button--variation Button--variation-tertiary"
      data-testid="template-inputs"
      type="button"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
        data-icon="template-inputs"
      >
        <svg
          height="28"
          viewBox="0 0 32 32"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M2.987 7.02a3 3 0 013-3h20.025a3 3 0 013 3v17.96a3 3 0 01-3 3H5.987a3 3 0 01-3-3v-5.638h2v5.639a1 1 0 001 1h20.025a1 1 0 001-1V7.02a1 1 0 00-1-1H5.987a1 1 0 00-1 1v5.673h-2V7.02z"
            fill="currentColor"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M12.184 12.446a1 1 0 011.399-.21l4.02 2.968a1 1 0 01.006 1.604l-4.02 3.016a1 1 0 11-1.2-1.6l1.62-1.216H2.988v-2h10.984l-1.576-1.164a1 1 0 01-.21-1.398z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        pipeline.templateInputs
      </span>
    </button>
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--withLeftIcon Button--variation Button--variation-tertiary"
      data-testid="input-variable"
      type="button"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
        data-icon="pipeline-variables"
      >
        <svg
          height="28"
          viewBox="0 0 20 20"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M2.197 4.143a.624.624 0 01.564-.356h7.489c.344 0 .624.28.624.624v1.872a.624.624 0 01-1.248 0V5.035H4.072l3.23 3.975c.178.22.186.53.021.76l-3.342 4.626h5.645v-1.248a.624.624 0 111.248 0v1.873c0 .344-.28.624-.624.624H2.76a.624.624 0 01-.505-.99L6.03 9.426 2.276 4.805a.624.624 0 01-.079-.662zM15.9 9.276a5.276 5.276 0 00-.724 1.032c-.22.44-.612.912-.994 1.288-.37.365-.824.732-1.205.884a.624.624 0 11-.463-1.159c.138-.055.445-.272.792-.614.336-.331.62-.692.753-.958.184-.366.529-.88.914-1.308.193-.215.414-.427.646-.59.218-.153.523-.319.87-.319a.624.624 0 01.012 1.248c0 .001-.052.013-.164.092-.129.09-.279.229-.437.404z"
            fill="current-color"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M13.334 9.276c.317.352.596.774.725 1.032.22.44.611.912.993 1.288.37.365.824.732 1.205.884a.624.624 0 10.464-1.159c-.139-.055-.445-.272-.793-.614-.336-.331-.62-.692-.753-.958a6.497 6.497 0 00-.913-1.308 3.98 3.98 0 00-.647-.59c-.217-.153-.523-.319-.87-.319a.624.624 0 00-.012 1.248c0 .001.052.013.165.092.128.09.278.229.436.404z"
            fill="current-color"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        variablesText
      </span>
    </button>
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--withLeftIcon Button--variation Button--variation-tertiary"
      type="button"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
        data-icon="pipeline-deploy"
      >
        <svg
          height="24"
          viewBox="0 0 16 16"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M14.8 1.148a.638.638 0 01.192.577l-2.396 13.707a.614.614 0 01-.43.49.584.584 0 01-.611-.193l-3.813-4.578-1.856 2.382a.584.584 0 01-.666.194.622.622 0 01-.398-.587V8.852l-4.46-2.32a.628.628 0 01-.329-.61.618.618 0 01.423-.543l13.772-4.362a.581.581 0 01.573.131zM6.02 8.783v2.6l4.996-6.413-4.997 3.813zm5.193-5.508L2.24 6.116 5.36 7.74l5.852-4.465zm2.182.62l-4.892 6.278a.59.59 0 01.065.067l3.074 3.69 1.753-10.035z"
            fill="current-color"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        notifications.pipelineName
      </span>
    </button>
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--withLeftIcon Button--variation Button--variation-tertiary"
      type="button"
    >
      <span
        class="bp3-icon bp3-icon-settings StyledProps--main StyledProps--padding-right-xsmall"
        icon="settings"
      >
        <svg
          data-icon="settings"
          height="20"
          viewBox="0 0 20 20"
          width="20"
        >
          <desc>
            settings
          </desc>
          <path
            d="M4 1c0-.55-.45-1-1-1S2 .45 2 1v5h2V1zM2 19c0 .55.45 1 1 1s1-.45 1-1v-6H2v6zm9-18c0-.55-.45-1-1-1S9 .45 9 1v8h2V1zm7 0c0-.55-.45-1-1-1s-1 .45-1 1v3h2V1zM9 19c0 .55.45 1 1 1s1-.45 1-1v-3H9v3zm9-14h-2c-.55 0-1 .45-1 1v3c0 .55.45 1 1 1h2c.55 0 1-.45 1-1V6c0-.55-.45-1-1-1zm-2 14c0 .55.45 1 1 1s1-.45 1-1v-8h-2v8zM4 7H2c-.55 0-1 .45-1 1v3c0 .55.45 1 1 1h2c.55 0 1-.45 1-1V8c0-.55-.45-1-1-1zm7 3H9c-.55 0-1 .45-1 1v3c0 .55.45 1 1 1h2c.55 0 1-.45 1-1v-3c0-.55-.45-1-1-1z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        pipeline.barriers.flowControl
      </span>
    </button>
    <button
      class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--without-shadow Button--withLeftIcon Button--variation Button--variation-tertiary"
      type="button"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
        data-icon="codebase-not-configured"
      >
        <svg
          height="20"
          viewBox="0 0 24 22"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M23.768 9.589l-5.577-5.9a.81.81 0 10-1.186 1.104l5.066 5.362-5.066 5.362a.808.808 0 00.31 1.311.81.81 0 00.876-.206l5.604-5.928a.808.808 0 00-.027-1.105zM6.954 3.634a.808.808 0 00-1.158.054L.218 9.59a.808.808 0 000 1.105l5.578 5.928a.81.81 0 001.185-1.105l-5.065-5.362L6.98 4.793a.808.808 0 00-.027-1.159zM14.436.02a.74.74 0 00-.569.094.777.777 0 00-.34.477L9.032 19.467a.8.8 0 00.09.584c.105.174.272.3.465.349h.177a.745.745 0 00.46-.158.78.78 0 00.273-.412L14.992.954A.795.795 0 0014.9.37a.76.76 0 00-.465-.35z"
            fill="#6B6D85"
          />
          <rect
            fill="#fff"
            height="11.6"
            rx="5.8"
            stroke="#FF832B"
            stroke-width="0.4"
            width="11.6"
            x="12.2"
            y="10.2"
          />
          <path
            clip-rule="evenodd"
            d="M18.384 12.223l3.558 6.231a.442.442 0 01-.384.661h-7.116a.442.442 0 01-.384-.66l3.558-6.232a.442.442 0 01.768 0zm-.602 4.866h.437l.224-2.833-.885-.006.224 2.839zm-.184 1.22v-.794h.805v.793h-.805z"
            fill="#FF832B"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        codebase
      </span>
    </button>
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main iconButton StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-size-xsmall Button--withLeftIcon Button--variation Button--variation-tertiary"
      type="button"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--padding-right-xsmall"
        data-icon="pipeline-advanced"
      >
        <svg
          height="24"
          viewBox="0 0 17 13"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M6.5 1H1v2.375h5.5V1zM1 0a1 1 0 00-1 1v2.375a1 1 0 001 1h5.5a1 1 0 001-1V1a1 1 0 00-1-1H1zm14.25 1h-5.5v2.375h5.5V1zm-5.5-1a1 1 0 00-1 1v2.375a1 1 0 001 1h5.5a1 1 0 001-1V1a1 1 0 00-1-1h-5.5z"
            fill="current-color"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M9.375 3h-2.5V2h2.5v1zm-5.125.917v3.096c0 .416.07.705.172.91.099.198.241.342.435.453.42.24 1.079.325 2.018.325v1c-.936 0-1.84-.072-2.514-.457a2.045 2.045 0 01-.834-.874c-.19-.381-.277-.835-.277-1.357V3.917h1zm9.135 2.417v-.43a.143.143 0 00-.143-.143h-.86a.143.143 0 00-.142.143v.964a.15.15 0 01-.107.14 2.273 2.273 0 00-.388.162.15.15 0 01-.175-.025l-.378-.378-.304-.303a.143.143 0 00-.202 0l-.608.607a.143.143 0 000 .203l.304.304.378.377a.15.15 0 01.024.176 2.283 2.283 0 00-.16.387.15.15 0 01-.142.107h-.964a.143.143 0 00-.143.143v.86c0 .078.064.143.143.143h.964a.15.15 0 01.141.106c.042.135.096.265.161.388a.15.15 0 01-.024.175l-.378.378-.304.304a.143.143 0 000 .202l.608.608a.143.143 0 00.202 0l.304-.304.378-.378a.15.15 0 01.175-.024c.123.065.253.119.388.16a.15.15 0 01.107.142v.964c0 .079.064.143.143.143h.86a.143.143 0 00.142-.143v-.964a.15.15 0 01.107-.141 2.28 2.28 0 00.388-.161.15.15 0 01.175.024l.378.378.304.304a.143.143 0 00.202 0l.608-.608a.143.143 0 000-.202l-.304-.304-.378-.378a.15.15 0 01-.024-.175 2.28 2.28 0 00.16-.388.15.15 0 01.142-.106h.964a.144.144 0 00.143-.144v-.859a.143.143 0 00-.143-.143h-.964a.15.15 0 01-.141-.107 2.283 2.283 0 00-.16-.387.15.15 0 01.023-.176l.378-.377.304-.304a.143.143 0 000-.203l-.608-.607a.143.143 0 00-.202 0l-.304.303-.378.378a.15.15 0 01-.175.025 2.273 2.273 0 00-.388-.161.15.15 0 01-.107-.141v-.534zm-.572 4.01a1.146 1.146 0 100-2.292 1.146 1.146 0 000 2.291zM8.75 9.697H7.5v-1h1.25v1z"
            fill="current-color"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        pipeline.advancedOptions
      </span>
    </button>
    <div />
  </div>
</div>
`;
